% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/align2reference.R
\name{.snpSharedPos}
\alias{.snpSharedPos}
\title{Find shared genomic position between two files.}
\usage{
.snpSharedPos(inputFile1, inputFile2, outputFile, nCore = 20)
}
\arguments{
\item{inputFile1}{the pure text file that has at least three columns:
chromosomal location, snp name and base-pair position, with the name 
c("chr", "rsID", "pos")}

\item{inputFile2}{the pure text file that has at least three columns:
chromosomal location, snp name and base-pair position, with the name 
c("chr", "rsID", "pos")}

\item{outputFile}{the pure text file return the snp name of 
the second input file.}

\item{nCore}{the number of cores used for computation.
The default is 20.}
}
\value{
The snp name of the second input file which shares the same genomic
position with that of the first input file.
}
\description{
Find shared genomic position between two files
and return the snp names of the second input file.
}
